### PR TITLE
dialogs' layouts fixes + FR polished

### PR DESCRIPTION
**Pull request by DRracer**
dialogs' layouts fixes (support menu dlgs, statistics etc.)
French translation polished

### CODE REVIEW

File: Firmware/menu.cpp [hunk 1]
```diff
@@ -423,7 +423,7 @@ void menu_draw_P<uint8_t*>(char chr, const char* str, int16_t val)
     float factor = 1.0f + static_cast<float>(val) / 1000.0f;
     if (val <= _md->minEditValue)
     {
-        lcd_printf_P(menu_fmt_float13off, chr, str, " [off]");
+        lcd_printf_P(menu_fmt_float13off, chr, str, _i(" [off]"));
```
Comment (mkbel) [inline Firmware/menu.cpp:426]: I don't see menu_fmt_float13off change in this commit to match change from string from RAM to PROGMEM.

Comment (mkbel) [reply Firmware/menu.cpp:426]: Fixed in:
14933f6b0e26894789af1284ad1b529f65a9e508

Fix. Sound [assist], Z-probe [5], menu_fmt_float13off format for last
string parameter from PROGMEM

File: Firmware/ultralcd.cpp [hunk 2]
```diff
@@ -1670,6 +1670,24 @@ static void lcd_cooldown()
   lcd_return_to_status();
 }
 
+//! @brief append text label with a colon and format it into a fixed size output buffer
+//! It would have been much easier if there was a ':' in the labels.
+//! But since the texts like Bed, Nozzle and PINDA are used in other places
+//! it is better to reuse these texts even though it requires some extra formatting code.
+static void pgmtext_with_colon(const char *ipgmLabel, char *dst, uint8_t dstSize){
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:1680]: Function parameters not documented.

File: Firmware/ultralcd.cpp [hunk 3]
```diff
@@ -1679,23 +1697,14 @@ void lcd_menu_extruder_info()                     // NOT static due to using ins
...
-	 _i("Nozzle FAN"),
-	 fan_speed_RPM[0],
-	 _i("Print FAN"),
-	 fan_speed_RPM[1]
-	);
+    // Display Nozzle fan RPM
+    lcd_timeoutToStatus.stop(); //infinite timeout
+    lcd_home();
+    char nozzle[12], print[12];
+    pgmtext_with_colon(_i("Nozzle FAN"), nozzle, 12);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp]: Do not use hardcoded constant if you can compute it. sizeof(nozzle)/sizeof(nozzle[0])

File: Firmware/ultralcd.cpp [hunk 4]
```diff
@@ -1679,23 +1697,14 @@ void lcd_menu_extruder_info()                     // NOT static due to using ins
...
+    pgmtext_with_colon(_i("Print FAN"), print, 12);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp]: Do not use hardcoded constant if you can compute it. sizeof(print)/sizeof(print[0])

File: Firmware/ultralcd.cpp [hunk 5]
```diff
@@ -1679,23 +1697,14 @@ void lcd_menu_extruder_info()                     // NOT static due to using ins
...
-	 ),
-	 _i("Nozzle FAN"),
-	 fan_speed_RPM[0],
-	 _i("Print FAN"),
-	 fan_speed_RPM[1]
-	);
+    // Display Nozzle fan RPM
+    lcd_timeoutToStatus.stop(); //infinite timeout
+    lcd_home();
+    char nozzle[12], print[12];
```
Comment (mkbel) [inline Firmware/ultralcd.cpp]: Declaring more variables in one line is not recommended especially for pointer types.

File: Firmware/ultralcd.cpp [hunk 6]
```diff
@@ -1877,16 +1887,23 @@ static void lcd_menu_debug()
 }
 #endif /* DEBUG_BUILD */
 
+//! @brief common line print for lcd_menu_temperatures
+static void lcd_menu_temperatures_line(const char *ipgmLabel, int value){
+    char tmp[15];
+    pgmtext_with_colon(ipgmLabel, tmp, 15);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp]: Do not use hardcoded constant if you can compute it. sizeof(tmp)/sizeof(tmp[0])

File: Firmware/ultralcd.cpp [hunk 7]
```diff
@@ -1670,6 +1670,24 @@ static void lcd_cooldown()
   lcd_return_to_status();
 }
 
+//! @brief append text label with a colon and format it into a fixed size output buffer
+//! It would have been much easier if there was a ':' in the labels.
+//! But since the texts like Bed, Nozzle and PINDA are used in other places
+//! it is better to reuse these texts even though it requires some extra formatting code.
+static void pgmtext_with_colon(const char *ipgmLabel, char *dst, uint8_t dstSize){
+    uint8_t i = 0;
+    for(; i < dstSize - 2; ++i){ // 1 byte less than buffer, we'd be adding a ':' to the end
```
Comment (mkbel) [inline Firmware/ultralcd.cpp]: Comment is misleading. Do you mean 2 bytes less than buffer, we'd be adding a ':' and '\0' ?

File: Firmware/ultralcd.cpp [hunk 8]
```diff
@@ -1670,6 +1670,24 @@ static void lcd_cooldown()
...
+        uint8_t b = pgm_read_byte(ipgmLabel + i);
+        if( ! b )
+            break;
+        dst[i] = b;
+    }
+    dst[i] = ':';               // append the colon
+    ++i;
+    for(; i < dstSize - 1; ++i) // fill the rest with spaces
+        dst[i] = ' ';
+    dst[dstSize-1] = 0;         // terminate the string properly
```
Comment (mkbel) [inline Firmware/ultralcd.cpp]: If you assign '\0', you can save the comment, as purpose  would be clear from code.

File: Firmware/ultralcd.cpp [hunk 9]
```diff
@@ -1745,12 +1754,12 @@ static void lcd_menu_fails_stats_mmu_print()
...
 //////////////////////
-	lcd_timeoutToStatus.stop(); //infinite timeout
+    lcd_timeoutToStatus.stop(); //infinite timeout
     uint8_t fails = eeprom_read_byte((uint8_t*)EEPROM_MMU_FAIL);
     uint16_t load_fails = eeprom_read_byte((uint8_t*)EEPROM_MMU_LOAD_FAIL);
-	lcd_home();
-	lcd_printf_P(PSTR("%S\n" " %S  %-3d\n" " %S  %-3d"), _i("Last print failures"), _i("MMU fails"), fails, _i("MMU load fails"), load_fails);
-	menu_back_if_clicked_fb();
+    lcd_home();
+    lcd_printf_P(PSTR("%S\n" " %-16.16S%-3d\n" " %-16.16S%-3d"), _i("Last print failures"), _i("MMU fails"), fails, _i("MMU load fails"), load_fails);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:1765]: Length hint for translators missing.

File: Firmware/ultralcd.cpp [hunk 10]
```diff
@@ -1779,14 +1789,14 @@ static void lcd_menu_fails_stats_total()
...
+    lcd_timeoutToStatus.stop(); //infinite timeout
     uint16_t power = eeprom_read_word((uint16_t*)EEPROM_POWER_COUNT_TOT);
     uint16_t filam = eeprom_read_word((uint16_t*)EEPROM_FERROR_COUNT_TOT);
     uint16_t crashX = eeprom_read_word((uint16_t*)EEPROM_CRASH_COUNT_X_TOT);
     uint16_t crashY = eeprom_read_word((uint16_t*)EEPROM_CRASH_COUNT_Y_TOT);
-	lcd_home();
-	lcd_printf_P(PSTR("%S\n" " %S  %-3d\n" " %S  %-3d\n" " %S  X %-3d  Y %-3d"), _i("Total failures"), _i("Power failures"), power, _i("Filam. runouts"), filam, _i("Crash"), crashX, crashY);
-	menu_back_if_clicked_fb();
+    lcd_home();
+    lcd_printf_P(failStatsFmt, _i("Total failures"), _i("Power failures"), power, _i("Filam. runouts"), filam, _i("Crash"), crashX, crashY);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:1802]: Length hint for translators missing.

File: Firmware/ultralcd.cpp [hunk 11]
```diff
@@ -1797,14 +1807,14 @@ static void lcd_menu_fails_stats_print()
...
+    lcd_timeoutToStatus.stop(); //infinite timeout
     uint8_t power = eeprom_read_byte((uint8_t*)EEPROM_POWER_COUNT);
     uint8_t filam = eeprom_read_byte((uint8_t*)EEPROM_FERROR_COUNT);
     uint8_t crashX = eeprom_read_byte((uint8_t*)EEPROM_CRASH_COUNT_X);
     uint8_t crashY = eeprom_read_byte((uint8_t*)EEPROM_CRASH_COUNT_Y);
-	lcd_home();
-	lcd_printf_P(PSTR("%S\n" " %S  %-3d\n" " %S  %-3d\n" " %S  X %-3d  Y %-3d"), _i("Last print failures"), _i("Power failures"), power, _i("Filam. runouts"), filam, _i("Crash"), crashX, crashY);
-	menu_back_if_clicked_fb();
+    lcd_home();
+    lcd_printf_P(failStatsFmt, _i("Last print failures"), _i("Power failures"), power, _i("Filam. runouts"), filam, _i("Crash"), crashX, crashY);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:1820]: Length hint for translators missing.

File: Firmware/ultralcd.cpp [hunk 12]
```diff
@@ -1877,16 +1887,23 @@ static void lcd_menu_debug()
 }
 #endif /* DEBUG_BUILD */
 
+//! @brief common line print for lcd_menu_temperatures
+static void lcd_menu_temperatures_line(const char *ipgmLabel, int value){
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:1897]: Function parameters not documented.

File: Firmware/ultralcd.cpp [hunk 13]
```diff
@@ -1877,16 +1887,23 @@ static void lcd_menu_debug()
...
+    lcd_timeoutToStatus.stop(); //infinite timeout
+    lcd_home();
+    lcd_menu_temperatures_line( _T(MSG_NOZZLE), (int)current_temperature[0] );
+    lcd_menu_temperatures_line( _T(MSG_BED), (int)current_temperature_bed );
 #ifdef AMBIENT_THERMISTOR
-	lcd_printf_P(PSTR(" %S:  %d%c\n" " PINDA:    %d%c"), _i("Ambient"), (int)current_temperature_ambient, '\x01', (int)current_temperature_pinda, '\x01');
-#else //AMBIENT_THERMISTOR
-	lcd_printf_P(PSTR(" PINDA:    %d%c"), (int)current_temperature_pinda, '\x01');
-#endif //AMBIENT_THERMISTOR
+    lcd_menu_temperatures_line( _i("Ambient"), (int)current_temperature_ambient );
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:1910]: Length hint for translators is missing.

File: Firmware/ultralcd.cpp [hunk 14]
```diff
@@ -1877,16 +1887,23 @@ static void lcd_menu_debug()
...
+#endif
+    lcd_menu_temperatures_line( _i("PINDA"), (int)current_temperature_pinda );
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:1912]: Length hint for translators is missing.

File: Firmware/ultralcd.cpp [hunk 15]
```diff
@@ -2659,7 +2676,7 @@ void lcd_menu_statistics()
 		lcd_clear();
 		lcd_printf_P(_N(
 			"%S:\n"
-			"%8.2fm\n"
+			"%17.2fm  \n"
 			"%S:\n"
 			"%2dh %02dm %02ds"
 		),_i("Filament used"), _met, _i("Print time"), _h, _m, _s);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:2689]: Length hint for translators is missing.

File: Firmware/ultralcd.cpp [hunk 16]
```diff
@@ -2681,7 +2698,7 @@ void lcd_menu_statistics()
 		lcd_clear();
 		lcd_printf_P(_N(
 			"%S:\n"
-			"%8.2fm\n"
+			"%17.2fm  \n"
 			"%S:\n"
 			"%7ldd :%2hhdh :%02hhdm"
 		), _i("Total filament"), _filament_m, _i("Total print time"), _days, _hours, _minutes);
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:2711]: Length hint for translators is missing.

File: Firmware/ultralcd.cpp [hunk 17]
```diff
@@ -2823,17 +2840,17 @@ static void lcd_menu_xyz_skew()
...
-	  "%S:  %5.2f\x01\n"
-	  "%S:  %5.2f\x01"
+	  "%-15.15S%3.2f\x01\n"
+	  "%-15.15S%3.2f\x01"
 	 ),
 	 _i("Measured skew"),
 	 separator,
-	 _i("Slight skew"), _deg(bed_skew_angle_mild),
-	 _i("Severe skew"), _deg(bed_skew_angle_extreme)
+	 _i("Slight skew:"), _deg(bed_skew_angle_mild),
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:2855]: Length hint for translators is missing.

File: Firmware/ultralcd.cpp [hunk 18]
```diff
@@ -2823,17 +2840,17 @@ static void lcd_menu_xyz_skew()
...
+	 _i("Severe skew:"), _deg(bed_skew_angle_extreme)
```
Comment (mkbel) [inline Firmware/ultralcd.cpp:2856]: Length hint for translators is missing.

**Review by mkbel — CHANGES_REQUESTED** [review]

**Comment by 3d-gussner** [comment]
@DRracer few of the comments from @mkbel could be solved if you pull the https://github.com/DRracer/Prusa-Firmware/tree/translations-doc as there are "Length hint for translators."

File: Firmware/ultralcd.cpp (hunk 3 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp]: @mkbel If you look into https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L1684 then you see that it has to do with placing the message correctly. As translations of a word or phrase can differ from English, i think it makes sense to "hardcode" it here.

File: Firmware/ultralcd.cpp (hunk 4 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp]: @mkbel See my comment above

File: Firmware/ultralcd.cpp (hunk 18 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:2856]: see https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L3012

File: Firmware/ultralcd.cpp (hunk 17 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:2855]: see https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L3011

File: Firmware/ultralcd.cpp (hunk 16 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:2711]: That one I also missed in https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L2858

File: Firmware/ultralcd.cpp (hunk 15 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:2689]: see https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L2836

File: Firmware/ultralcd.cpp (hunk 14 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:1912]: This have been documented in https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L1961 . just needs a merge.

File: Firmware/ultralcd.cpp (hunk 6 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp]: See my comment above

File: Firmware/ultralcd.cpp (hunk 9 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:1765]: See https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L1780

File: Firmware/ultralcd.cpp (hunk 10 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:1802]: See https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L1826

File: Firmware/ultralcd.cpp (hunk 11 above)
Comment (3d-gussner) [reply Firmware/ultralcd.cpp:1820]: See https://github.com/DRracer/Prusa-Firmware/blob/translations-doc/Firmware/ultralcd.cpp#L1849

File: Firmware/ultralcd.cpp (hunk 3 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp]: @mkbel I'd rather define a static const int something than doing sizeof(nozzle)/sizeof(nozzle[0]), but you are right. Hardcoding 3x the same constant isn't a final code. @3d-gussner is correct, the input text will be cut to 11 characters max no matter what comes from the translation engine

File: Firmware/ultralcd.cpp (hunk 16 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp:2711]: @3d-gussner we definitely shall merge the translation-doc ASAP, there is so much work that will improve this.

File: Firmware/ultralcd.cpp (hunk 2 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp:1680]: @mkbel will be added

File: Firmware/ultralcd.cpp (hunk 3 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp]: @mkbel will be fixed in next commit

File: Firmware/ultralcd.cpp (hunk 5 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp]: @mkbel in this case it is reasonably readable

File: Firmware/ultralcd.cpp (hunk 6 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp]: @mkbel will be fixed in next commit

File: Firmware/ultralcd.cpp (hunk 7 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp]: @mkbel will be fixed in next commit

File: Firmware/ultralcd.cpp (hunk 8 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp]: @mkbel will be fixed in next commit

File: Firmware/ultralcd.cpp (hunk 12 above)
Comment (DRracer) [reply Firmware/ultralcd.cpp:1897]: @mkbel will be fixed in next commit

**Review by mkbel — APPROVED** [review]

**Pull request merged**